자기소개서 기능 개발

Kongji82 opened the pull request:
# Pull Request 요약

- 자기소개서 기능 개발 

## 변경 사항

### 자기소개서 생성
- 자기소개서 등록 구현
- 제목, 내용 작성 가능
- 헤더에 Authorization을 통해 작성자 확인
- 엔드포인트: POST api/v2/resume/create

### 자기소개서 삭제
- 자기소개서 삭제 기능
- 헤더에 Authorization을 통해 작성자 확인
    - 자기소개서 작성자가 아닌 경우 예외처리
- 엔드포인트: DELETE api/v2/resume/{resumeId}
  
### 자기소개서 편집
- 자기소개서 편집 기능
- 헤더에 Authorization을 통해 작성자 확인
    - 자기소개서 작성자가 아닌 경우 예외처리
- 엔드포인트: POST api/v2/resume/{resumeId}

### 자기소개서 가져오기
- 자기소개서를 지정된 resumeId에 따라 가져오기
- 엔드포인트: GET api/v2/resume/{resumeId}

## 개발 유형

- [x] 새로운 기능 개발
- [x] 테스트 코드 작성

## 작업 기간

- 작업 시작일: (2024-03-06)
- 작업 종료일: (2024-03-16)

Kongji82 committed:
[feat] Resume 생성 테스트코드 작성

변경 사항:
- Resume 생성 테스트 코드 작성

Kongji82 committed:
[feat] Resume Entity 구현

변경 사항:
- Resume Entity 구현

Kongji82 committed:
[feat] Resume 생성 기능 구현

변경 사항:
- Resume 생성 기능 구현

Kongji82 committed:
[refactor] ResumeServiceTest 리팩토링

변경 사항:
- Test에 자주 사용되는 Resume Mock 메소드 추출
- 변수 추출

Kongji82 committed:
[test] Resume update 테스트 코드 작성

변경 사항:
- Resume Update 테스트 코드 작성

Kongji82 committed:
[feat] Resume Update 기능 추가

변경 사항:
- Resume Update 기능 추가
- ErrorCode에 RESUEM_NOT_FOUND, HAS_NO_PERMISSION_WITH_THIS_RESUME Exception 추가
-

Kongji82 committed:
[test] updateResume() 에러 테스트코드 작성

- 자기소개서 작성자가 아닌 경우 NoResumeWriterException

Kongji82 committed:
[test] getResume 테스트 코드 작성

- 정상적인 작업 & 없는 id를 찾는 경우

Kongji82 committed:
[feat] getResume api 구현

Kongji82 committed:
[refactor] Resume viewCount 추가

- rating column -> likeCount 이름  변경
- 조회수 viewCount 추가

Nine-JH commented on src/main/java/com/fasttime/domain/resume/controller/ResumeController.java:
```diff
@@ -0,0 +1,82 @@
+    @PutMapping("/{resumeId}")
+    public ResponseEntity<ResponseDTO<ResumeResponseDto>> updateResume(
+            @PathVariable Long resumeId,
+            @RequestBody @Valid ResumeUpdateRequest request
```
해당 DTO는 어떤 인증값을 검증하기 위해서 `@Valid`을 사용하신 건가요? 해당 dto에 검증을 하는 코드가 따로 보이지는 않아서요
추가적으로 update 쪽 부분 뿐 아니라 create나 다른 곳도 검증을 같이 해나갔으면 좋아보이네요

Kongji82 replied:
`@Valid` 같은 경우 update하기 위해 null값이 안넘어오도록 체크하기위해 작성했습니다.
create도 그렇게 수정하도록 하는게 좋을것 같네요! 수정하겠습니다.

Nine-JH commented on src/main/java/com/fasttime/domain/resume/service/ResumeService.java:
```diff
@@ -0,0 +1,95 @@
+        isWriter(requestMember, resume);
+
+        resume.updateResume(request.title(), request.content());
+        resumeRepository.save(resume);
```
`save()`를 명시적으로 사용하신 이유가 있으실까요??

Kongji82 replied:
update에 대해 entity에서 값을 업데이트 하더라도 JPA에 적용이 안된다 생각해서 `save()`를 사용하였습니다.

Nine-JH replied:
네 아마 JPA의 `더티 체킹`에 관한 기능덕분에 따로 `save()`를 사용하지 하지 않아도 저장이 될 거예요! 
또 만약 JPA가 아닌 다른 repository 관련 의존성을 사용할 가능성도 염두해서 `save()`를 썻다면 그것대로도 일리가 있는 의견이기 때문에 곱씹어보시면 좋을 것 같아요~

Nine-JH commented on src/main/java/com/fasttime/domain/resume/entity/Resume.java:
```diff
@@ -0,0 +1,65 @@
+        this.likeCount += 1;
+    }
+
+    public void view(){
+        this.viewCount += 1;
+    }
```
조회수를 집계하는 기능도 계획하시는군요!
꽤나 어려운 설계가 될 겁니다. 새로고침과 같은 비정상적인 방법으로 조회수를 늘리는 방법도 막아야 하고, 같은 계정이나 IP로 조회를 했을 때 하루에 1회 조회수를 늘려줄것인가 또는 딱 한번만 늘려줄것인가 등등.. 좋은 비즈니스 로직 기대하겠습니다 ㅎㅎ

Nine-JH reviewed (commented):
고생하셨습니다! 
앞으로도 잘 부탁드립니다~

Kongji82 committed:
[refactor] Resume Create() 유효성 검사 추가

Kongji82 committed:
[refactor] Resume create() & update() Dto 유효성 검사 로직 추가

- ResumeUpdateRequest @NotNull 유효성 검사 추가
- ResumeRequestDto @NotNull 유효성 검사 추가

ghrltjdtprbs commented on src/main/java/com/fasttime/domain/resume/controller/ResumeController.java:
```diff
@@ -0,0 +1,82 @@
+    private final ResumeService resumeService;
+
+    // Resume create
+    @PostMapping("/create")
```
RESTful API 설계에서는 일반적으로 URL이 리소스를 나타내고, HTTP 메소드가 해당 리소스에 대한 행동을 명시합니다.
따라서 POST /api/v2/resume/create 보다는 
POST /api/v2/resumes이 더 RESTful 할 것 같습니다! 참고해주세요! 😊😊

ghrltjdtprbs commented on src/main/java/com/fasttime/domain/resume/dto/ResumeDeleteServiceRequest.java:
```diff
@@ -0,0 +1,12 @@
+public record ResumeDeleteServiceRequest(
+        Long resumeId,
+        Long requestUserId,
+        LocalDateTime deleteAt
```
해당 deleteAt이 어떻게 사용되는지 궁금합니다! soft delete를 적용하기 위해서라면 엔터티에서 상속 받은 BaseTimeEntity를 이용하면 되지 않나요??

ghrltjdtprbs commented on src/main/java/com/fasttime/domain/resume/service/ResumeService.java:
클래스 레벨에 @Transactional 추가하는 것은 어떤가요?? 
이미 아시겠지만, 데이터 일관성을 보장하고, getResume같은 단순 조회 메소드에는 읽기 전용 트랜잭션으로 성능 향상도 노려볼 수 있습니다! 👍

ghrltjdtprbs commented on src/test/java/com/fasttime/domain/resume/unit/service/ResumeServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+                .writer(member)
+                .build();
+    }
+}
```
코드 스타일을 통일해보는 것은 어떨까요???
1. 메서드 선언의 맨 위와 아래는 줄바꿈 X
2. 클래스 선언의 맨 위는 줄바꿈 O, 맨 아래는 줄바꿈 X
3. 파일의 맨 아래에 줄바꿈 삽입 (맨 아래 개행이 필요한 이유: https://seongwon.dev/Git/20220303-%ED%8C%8C%EC%9D%BC%EC%9D%98_%EB%A7%88%EC%A7%80%EB%A7%89_%EA%B0%9C%ED%96%89/)

ghrltjdtprbs reviewed (approved):
수고하셨습니다! 간단한 코멘트와 질문 남겼으니 확인 부탁드립니다 ㅎㅎ 😊😊

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/controller/ResumeController.java:
```diff
@@ -0,0 +1,82 @@
+
+    // Resume create
+    @PostMapping("/create")
+    public ResponseEntity<ResponseDTO<ResumeResponseDto>> creatResume(
```
오타가 있는 것 같아요~ creatResume -> createResume
```suggestion
    public ResponseEntity<ResponseDTO<ResumeResponseDto>> createResume(
```

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/controller/ResumeController.java:
```diff
@@ -0,0 +1,82 @@
+            @PathVariable Long resumeId,
+            @RequestBody @Valid ResumeUpdateRequest request
+    ) {
+
```
```suggestion
```
공백 라인이 필요 없을 것 같네요!

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/dto/ResumeResponseDto.java:
```diff
@@ -0,0 +1,15 @@
+        String writer,
+        int likeCount,
+
+        int viewCount) {
```
```suggestion
        int viewCount
) {
```
수정해주시면 좀 더 일관성 있는 코드가 될 것 같습니다~

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/dto/ResumeUpdateServiceRequest.java:
```diff
@@ -0,0 +1,9 @@
+        Long resumeId,
+        Long memberId,
+        String title,
+        String content) {
```
```suggestion
        String content
) {
```
위 코멘트와 마찬가지 입니다!

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/entity/Resume.java:
```diff
@@ -0,0 +1,65 @@
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Builder
```
@AllArgsConstructor는 필드 위치를 변경했을 때 생성자의 파라미터 순서를 변경시킵니다. 이 경우에는 IDE가 제공해주는 리팩토링이 작동하지 않기도 하고, 동일한 타입의 파라미터를 사용할 경우 오류도 발생하지 않아 인자 값이 바뀌어 들어가게 되는 버그를 인식하기가 어려습니다. 

생성자에 @Builder를 붙이면, 파라미터 순서가 아닌 인자 이름 인식하기 때문에 위에 말씀 드린 버그를 방지할 수 있습니다. 

때문에 생성자를 직접 구현하고 생성자에 @Builder를 붙이는 방법을 추천 드립니다! 😊

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/entity/Resume.java:
```diff
@@ -0,0 +1,65 @@
+    private String title;
+    private String content;
+    @ColumnDefault("0")
+    private int likeCount;
```
필드 간 줄 띄움을 맞추면 더 좋을 것 같습니다 ㅎㅎ
```suggestion
    private String title;

    private String content;

    @ColumnDefault("0")
    private int likeCount;
```

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/entity/Resume.java:
```diff
@@ -0,0 +1,65 @@
+    @ManyToOne
+    private Member writer;
+
+    public static Resume createNewResume(Member member, String title, String content) {
+        return Resume.builder()
+                .title(title)
+                .content(content)
+                .writer(member)
+                .build();
+    }
```
빌더를 사용했을 때 인자 이름으로 인식하여 파라미터 순서를 헷갈리지 않을 수 있다는 장점이 가려지는 것 같습니다. 같은 자료형을 쓰고 있는 title, content도 있는 만큼, 휴먼 에러를 유발할 수 있을 것 같다는 우려가 드는데, Resume 생성에 빌더를 사용하지 않고 따로 스태틱 메서드를 정의하신 이유가 있을까요?

Kongji82 replied:
Resume에 대한 생성에 대한 책임을 Entity에 주고 싶어서 만들었습니다. 하지만, 빌더를 사용했으니까 따로 만들어 쓸 필요가 없겠군요. 
수정하도록 하겠습니다!

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/entity/Resume.java:
```diff
@@ -0,0 +1,65 @@
+    private String title;
+    private String content;
+    @ColumnDefault("0")
+    private int likeCount;
```
likeCount가 Resume 테이블의 속성으로 들어가 있네요! 
이 설계의 경우 한 회원이 계속해서 좋아요를 누를 수 있을 것 같습니다~ 
또, 좋아요를 취소하는 것도 어려울 것 같네요~
한 회원은 좋아요를 한 번 누를 수 있고, 해당 좋아요를 취소할 수도 있도록 재설계 해보시는 것은 어떨까요?

Kongji82 replied:
이 부분은 나중에 좋아요 기능 설계하면서 바꿔 보도록 하겠습니다.

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/exception/NoResumeWriterException.java:
```diff
@@ -0,0 +1,20 @@
+import com.fasttime.global.exception.ApplicationException;
+import com.fasttime.global.exception.ErrorCode;
+
+public class NoResumeWriterException extends ApplicationException {
+    private static final ErrorCode ERROR_CODE = ErrorCode.HAS_NO_PERMISSION_WITH_THIS_RESUME;
```
구글 스타일에 맞게 공백 추가해주세요~
```suggestion
public class NoResumeWriterException extends ApplicationException {

    private static final ErrorCode ERROR_CODE = ErrorCode.HAS_NO_PERMISSION_WITH_THIS_RESUME;
```

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/service/ResumeService.java:
```diff
@@ -0,0 +1,95 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Slf4j
```
로깅 기능을 사용하지 않으신 것 같아, 해당 애노테이션은 빼도 될 것 같습니다!

JeongUijeong commented on src/main/java/com/fasttime/domain/resume/service/ResumeService.java:
```diff
@@ -0,0 +1,95 @@
+
+    private void isWriter(Member requestMember, Resume resume) {
+        if (!requestMember.getId().equals(resume.getWriter().getId())) {
+            throw new NoResumeWriterException("");
```
message를 따로 설정할 수 있는 생성자를 사용하셨는데, 
여기서는 기본 생성자로 생성하셔도 ErrorCode에 명시되어 있는 "해당 자기소게서에 대한 권한이 없습니다."라는 메시지를 넘길 수 있을 것 같습니다!

빈 스트링을 파라미터로 넘기신 이유가 있나요?

JeongUijeong commented on src/test/java/com/fasttime/domain/resume/unit/entity/ResumeTest.java:
```diff
@@ -0,0 +1,37 @@
+package com.fasttime.domain.resume.unit.entity;
+
+
```
```suggestion
```
공백 라인 삭제해주세요~

JeongUijeong commented on src/test/java/com/fasttime/domain/resume/unit/entity/ResumeTest.java:
```diff
@@ -0,0 +1,37 @@
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+
+@Slf4j
```
로깅 사용하지 않으신다면 해당 애노테이션 삭제하셔도 될 것 같습니다!
```suggestion
```

JeongUijeong commented on src/test/java/com/fasttime/domain/resume/unit/service/ResumeServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+            given(resumeRepository.findById(MOCK_RESUME_ID)).willReturn(Optional.of(mockResume));
+
+            // when
+            ResumeResponseDto response = resumeService.updateResume(updateRequest);
+            // then
```
별 거 아니지만 공백 라인 하나가 누락된 것 같습니다~
```suggestion
            ResumeResponseDto response = resumeService.updateResume(updateRequest);

            // then
```

JeongUijeong commented on src/test/java/com/fasttime/domain/resume/unit/service/ResumeServiceTest.java:
// given

// when

// then

주석을 달아 명시적으로 구분해주신다면, 모든 테스트 메서드에 달아주시면 좋을 것 같습니다~
resume_validateFail_throwIllegalArgumentException 메서드 이후부터는 주석이 없네요!

JeongUijeong reviewed (changes requested):
수고 많으셨습니다~👍🏻👍🏻
간단한 코드 스타일 관련 코멘트와 제안, 코드에 대한 질문 코멘트를 몇 개 남겼습니다!
바쁘시겠지만, 확인 부탁드려요!😊

Kongji82 committed:
[refactor] URL 형식 변경

- "/api/v2/resume" -> "/api/v2/resumes"

Kongji82 committed:
[refactor] Delete 로직 변경

- DeleteAt을 request값에 처리하는것이 아닌 Server Service Layer에서 처리

Kongji82 committed:
[refactor] Convention에 맞게 코드 수정

Kongji82 committed:
[refactor] Resume Builder 패턴 수정

- @AllArgConstructor 삭제
- 생성자 작성, @Builder 적용
- ResumeService createResume() 로직 수정

Kongji82 committed:
[refactor] 작성 권한이 없는 유저에 대한 오류 메시지 작성

- isWriter() 권한이 없는 경우 오류 메시지 작성

Kongji82 committed:
[refactor] 공백문자 제거

- ResumeTest 상단 공백문자 제거

Kongji82 committed:
[bug-fix] Resume 컴파일시 DB에 Create

- DB Resume DDL 추가

Kongji82 committed:
Merge branch 'develop' into feature/resume

ghrltjdtprbs reviewed (approved):
반영사항 확인했습니다! 수고하셨습니다 ㅎㅎ👍
